theory/DD/countpairs_impl.c.src

```diff
@@ -228,9 +228,22 @@ int countpairs_DOUBLE(const int64_t ND1, DOUBLE *X1, DOUBLE *Y1, DOUBLE *Z1,
             options->boxsize);
         return EXIT_FAILURE;
     }
-    const DOUBLE xdiff = options->boxsize > 0 ? options->boxsize:(xmax-xmin);
-    const DOUBLE ydiff = options->boxsize > 0 ? options->boxsize:(ymax-ymin);
-    const DOUBLE zdiff = options->boxsize > 0 ? options->boxsize:(zmax-zmin);
+
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wfloat-equal"
+#endif
+    
+    double boxsize_y = options->boxsize_y == -1. ? options->boxsize : options->boxsize_y;
```

Comment:
Not sure I follow the intent here. Why would the user specify -1 for any specific direction if they wanted the overall box size - would be faster to either type the boxsize, or in case of a cube, pass the scalar boxsize. 

Wouldn't it make more sense if -1 was implying non-periodic along specific axes?

Reply:
Or may be this is a different feature - which would have to be enabled by accepting a triple for `periodic`. For example, `(1, 0, 1)` would imply periodic X & Y but not Z. However, I have a feeling that adding such a feature will make the code too complicated

Reply:
`-1` means "inherit from the scalar boxsize" and is the default (so as not to change any existing code).

It would be nice to accept a triple for `periodic` as well, but I agree it might be too complicated for now. At least with this PR, one can mock that behavior by setting the boxsize in a given dimension to be much larger than the particle distribution!

Reply:
Not sure about the necessity of the "-1" behaviour. From the user perspective, if they want to use the same boxsize along two dimensions, then typing that value in once more is fine. 

This is do-able without accepting a triplet for periodic. If *any* of the dimensions are periodic, then the user has to specify `periodic=True`, and then must provide either a scalar or a triplet for boxsize. Consider the following scenarios:

- user provides a scalar/1-element boxsize, then that (scalar) boxsize is broadcast to all three dimensions and all of the dims are periodic

(*all following cases require 3 values in boxsize*)
- user provides three different (non-zero, positive) values of the boxsize, then each dim has those separate periodic wrapping values
- user provides 0.0 along any dim, then auto-detect kicks in with the particle extents
- user provides any negative value, periodicity is turned off along those dims -> which effectively means that the wrapping value is set to 0.0


Two related checks come to mind:
- Do we check that the user-provided boxsize (along any dimension) is bigger than the particle extent?
- How do we add the check for a single-item list/tuple for boxsize? `np.isscalar([boxsize])` will return `False`

Reply:
That sounds like a good user experience to me. The value of `-1` wasn't intended to be user-facing; rather, it's just the default value in the `config` struct, so we know whether the user code set that field or not. We can change `-1` to mean "not periodic" and use `-2` as the "not set by user" value instead. This is particularly important for the C API, so we know whether the client code set `boxsize_y` and `boxsize_z` or not.

> user provides any negative value, periodicity is turned off along those dims -> which effectively means that the wrapping value is set to 0.0

Looking at the code, I think the place to make this change is [`generate_cell_pairs_DOUBLE()`](https://github.com/manodeep/Corrfunc/blob/4e5e9bcd4b10d856f134778de2f41b138cbd022e/utils/gridlink_impl.c.src#L412).  That function takes `int periodic` and `DOUBLE x/y/zdiff`.  Since `x/y/zdiff` is always positive, we can use `-1` to signal a non-periodic dimension, and `int periodic` will mean "one or more dim might be periodic" (or we can just get rid of that var and look at `x/y/zdiff`).  So like:
```c
generate_cell_pairs_DOUBLE(...,
    const DOUBLE xdiff, const DOUBLE ydiff, const DOUBLE zdiff,
...) {
    int periodic_x = xdiff >= 0;
    int periodic_y = ydiff >= 0;
    int periodic_x = zdiff >= 0;
...
}
```

Does that seem reasonable? Anywhere else you can think of that will have to change?

Reply:
<snip>
 
> Two related checks come to mind:
> 
>     * Do we check that the user-provided boxsize (along any dimension) is bigger than the particle extent?
> 
</snip>

Answering my question - yes, Corrfunc does make that check https://github.com/manodeep/Corrfunc/blob/4e5e9bcd4b10d856f134778de2f41b138cbd022e/utils/gridlink_impl.c.src#L147 and 
https://github.com/manodeep/Corrfunc/blob/4e5e9bcd4b10d856f134778de2f41b138cbd022e/utils/gridlink_impl.c.src#L156

Reply:
> Looking at the code, I think the place to make this change is [`generate_cell_pairs_DOUBLE()`](https://github.com/manodeep/Corrfunc/blob/4e5e9bcd4b10d856f134778de2f41b138cbd022e/utils/gridlink_impl.c.src#L412). That function takes `int periodic` and `DOUBLE x/y/zdiff`. Since `x/y/zdiff` is always positive, we can use `-1` to signal a non-periodic dimension, and `int periodic` will mean "one or more dim might be periodic" (or we can just get rid of that var and look at `x/y/zdiff`). 

Probably better to deal with as an enum rather than hard-coding values. 

> So like:
> 
> ```c
> generate_cell_pairs_DOUBLE(...,
>     const DOUBLE xdiff, const DOUBLE ydiff, const DOUBLE zdiff,
> ...) {
>     int periodic_x = xdiff >= 0;
>     int periodic_y = ydiff >= 0;
>     int periodic_x = zdiff >= 0;
> ...
> }
> ```
> 
> Does that seem reasonable? Anywhere else you can think of that will have to change?

Need to do a search and think through that duplicate cells + min-sep.